csrc/utils.h

```diff
@@ -621,86 +623,205 @@ void checkAllEqual(std::initializer_list<T> elements) {
 
 #if __cplusplus >= 202302L
 
+using std::views::enumerate;
 using std::views::zip;
 
 #else
 
 namespace views {
-
-template <std::ranges::input_range... Rs>
+template <std::ranges::view... Rs>
 class zip_view : public std::ranges::view_interface<zip_view<Rs...>> {
  private:
   std::tuple<Rs...> bases;
 
-  // Iterator for begin()
-  struct begin_iterator {
+  static constexpr bool is_bidirectional =
+      (std::ranges::bidirectional_range<Rs> && ...);
+
+  struct iterator {
     std::tuple<std::ranges::iterator_t<Rs>...> iterators;
 
     using value_type = std::tuple<std::ranges::range_value_t<Rs>...>;
     using reference = std::tuple<std::ranges::range_reference_t<Rs>...>;
     using difference_type = std::ptrdiff_t;
+    using iterator_category = std::conditional_t<
+        is_bidirectional,
+        std::bidirectional_iterator_tag,
+        std::input_iterator_tag>;
 
-    begin_iterator& operator++() {
+    iterator& operator++() {
       std::apply([](auto&... it) { ((++it), ...); }, iterators);
       return *this;
     }
 
+    iterator operator++(int) {
+      iterator temp = *this;
+      ++(*this);
+      return temp;
+    }
+
+    // Enable bidirectional iteration only if `is_bidirectional` is true
+    iterator& operator--() requires is_bidirectional {
+      std::apply([](auto&... it) { ((--it), ...); }, iterators);
+      return *this;
+    }
+
+    iterator operator--(int) requires is_bidirectional {
+      iterator temp = *this;
+      --(*this);
+      return temp;
+    }
+
     reference operator*() const {
       return std::apply(
           [](auto&... it) -> reference { return {*it...}; }, iterators);
     }
 
-    bool operator==(const begin_iterator& other) const {
-      return iterators == other.iterators;
-    }
+    bool operator==(const iterator& other) const = default;
   };
 
-  // Sentinel for end()
-  struct end_iterator {
+  struct sentinel {
     std::tuple<std::ranges::sentinel_t<Rs>...> sentinels;
 
-    bool operator==(const begin_iterator& it) const {
+    bool operator==(const iterator& it) const {
       return compare(it, std::make_index_sequence<sizeof...(Rs)>{});
     }
 
    private:
     template <std::size_t... I>
-    bool compare(const begin_iterator& it, std::index_sequence<I...>) const {
+    bool compare(const iterator& it, std::index_sequence<I...>) const {
       return ((std::get<I>(it.iterators) == std::get<I>(sentinels)) || ...);
     }
   };
 
  public:
-  explicit zip_view(Rs&&... ranges)
-      : bases{std::forward<Rs>(ranges)...} {} // Ensure perfect forwarding
+  zip_view() = default;
```

Comment:
Why is this needed now?

Reply:
C++23 standard does define such a ctor: https://en.cppreference.com/w/cpp/ranges/zip_view/zip_view, so I just added it. But I doubt that it would be interesting for us, and I can not think of an interesting case to test on.